056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+- From the quorum info returned from the admin API, fetch `lastCaughtUpTimestamp` for every controller except the current controller we are considering to roll. `lastCaughtUpTimestamp` is the last millisecond timestamp at which a replica controller was known to be caught up with the quorum leader.
+- Check the quorum leader id using the quorum info and identify the `lastCaughtUpTimestamp` of the quorum leader
+- Retrieve value of the Kafka property `controller.quorum.fetch.timeout.ms` from the desired configurations specified in the Kafka CR. If this property does not exist in the desired configurations, then use the hard-coded default value for it which is `2000`. The reason for this is explained further in the **NOTE** below.
+- Mark a node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`.
+- Count each controller node that is caught up, including the leader (`numOfCaughtUpControllers`).
+- Can roll if: `numOfCaughtUpControllers > ceil((double) totalNumOfControllers / 2)`.
+
+> NOTE: Until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) is implemented, KafkaRoller cannot create an admin connection to the controller directly to describe its configuration or the quorum state. Therefore KafkaRoller checks the desired configurations to get the value of `controller.quorum.fetch.timeout.ms` and creates admin client connection to the brokers for the quorum check. If KafkaRoller cannot connect to any of the brokers after the maximum attempts to retry, the controllers will be marked as failed to reconcile because of not being able to determine the quorum health. KafkaRoller would then try to reconcile the brokers, which may help restoring admin client connections to them. In this scenario, the reconcilation will be completed with failure and reported to the operator. The controllers quorum check will be retried in the next round of reconcilation.
+
+
+#### Separate brokers and controllers
+For controller-only:
+ 1. Does not force roll a controller performing log recovery
+ 2. Force rolls a broker if the pod is in stuck state
+ 3. Attempts to connect an admin client to any of the brokers, if cannot connect, skip checks in `4` and `5` 
+ 4. Rolls a controller if the controller configuration has changed
+ 5. Does not roll a controller if doing so would take the number of caught-up controllers (inc leader) to less than half of the quorum
+
+For broker-only:
+ 1. Does not force roll a broker performing log recovery
+ 2. Force rolls a broker if the pod is in stuck state
+ 3. **NEW** Attempts to connect an admin client to a broker and if it can't connect rolls the broker 
+ (the order of the rollability checks is different, we only attempt admin client connection and force roll if the pod is not stuck)
```

Comment:
Could you explain how we do `only attempt admin client connection`?

Reply:
Currently KafkaRoller checks whether admin client is created successfully and if not, it immediately force rolls the broker. This sentence is trying to say, we execute this step only if pod is not stuck. I will reword it little bit better.